src/abstract-api-adapter.coffee

```diff
@@ -0,0 +1,119 @@
+		* @return {string} Next page url
+		###
+		if not @extractNext?
+			throw new TypeError('Must implement extractNext')
+		###*
+		* Given the parsed body from the Rest API, extract new ids object and pass it along the chain.
+		* You can assume, by this stage, that the HTTP request returned error is falsey and statusCode is 200.
+		* @param {object} Objectified body from the HTTP request
+		* @param {function} (error, ids) next function in the chain.  ids = {thread, comment?}
+		###
+		if not @parseResponse?
+			throw new TypeError('Must implement parseResponse')
+		###*
+		* Given suitable details will return request details
+		* @param {string} key of the identity to use
+		* @param {string} id of the flow to update
+		* @param {string} text to post
+		* @param {string}? id of the thread to update
+		* @return {object} {url, headers, payload}
+		###
+		if not @buildRequest?
+			throw new TypeError('Must implement buildRequest')
+		###*
+		* Triggers a poll of the API
+		* Likely to use getUntil
+		###
+		if not @poll?
+			throw new TypeError('Must implement poll')
+		super
+		@interval = 1000
+		@lastReport = null
+		@pollInProgress = true
+
+	run: ->
+		setInterval @maybePoll, @interval
+		@emit 'connected'
+
+	maybePoll: =>
+		unless @pollInProgress
+			@pollInProgress = true
+			@poll()
+
+	# Execute a series of paged requests until we run out of pages or a filter rejects
+	getUntil: (options, each, filter = -> true) -> new Promise (resolve, reject) =>
+		request.get? options, (error, response, body) =>
+			if error or response?.statusCode isnt 200
+				if response?.statusCode in [429, 503]
+					@report { error: error, html: response?.statusCode, url: options.url }
+					# Retry if the error code indicates temporary outage
+					@getUntil(options, each, filter)(resolve, reject)
+				else
+					@robot.logger.error response?.statusCode + '-' + options.url
+					@finishPoll { error: error, html: response?.statusCode, url: options.url }
+					reject(error ? new Error("StatusCode: #{response.statusCode}"))
+			else
+				responseObject = JSON.parse(body)
+				results = @extractResults responseObject
+				for result in results
+					if filter result
+						each result
+					else
+						@finishPoll { html: response?.statusCode, url: options.url }
+						resolve('Reached filtered object')
+						return
+				if @extractNext responseObject
+					options.url = @extractNext responseObject
+					@getUntil(options, each, filter)(resolve, reject)
+				else
+					@finishPoll { html: response?.statusCode, url: options.url }
+					resolve('Reached end of pagination')
+
+	finishPoll: (report) ->
+		@pollInProgress = false
+		report = JSON.stringify(report)
+		if @lastReport isnt report
+			@lastReport = report
+			@robot.logger.debug report
+
+	###*
+	* Given a set of ids make best effort to publish the text and pass on the published ids
+	* @param {string} text to publish
+	* @param {object} ids to use.  ids = {user, flow, thread?}
```

Comment:
Every one of our chat systems uses a different terminology to mean the same thing.  It's muddling me up so much that I've got a chart of the equivalences in my exercise book.
This said, the most common comms tool in resin is Flowdock, so where I've a choice I've used Flowdock's nouns.